libethereum/WarpPeerCapability.cpp

```diff
@@ -0,0 +1,175 @@
+	Capability(_s, _h, _i, _capID)
+{
+}
+
+void WarpPeerCapability::init(unsigned _hostProtocolVersion, u256 _hostNetworkId, u256 _chainTotalDifficulty, h256 _chainCurrentHash, h256 _chainGenesisHash, std::shared_ptr<WarpPeerObserverFace> _observer)
+{
+	m_hostProtocolVersion = _hostProtocolVersion;
+	m_observer = _observer;
+	requestStatus(_hostNetworkId, _chainTotalDifficulty, _chainCurrentHash, _chainGenesisHash);
+}
+
+bool WarpPeerCapability::interpret(unsigned _id, RLP const& _r)
+{
+	m_lastAsk = std::chrono::system_clock::to_time_t(chrono::system_clock::now());
+	try
+	{
+		switch (_id)
+		{
+		case WarpStatusPacket:
+		{
+			// Packet layout:
+			// [ version:P, state_hashes : [hash_1:B_32, hash_2 : B_32, ...],  block_hashes : [hash_1:B_32, hash_2 : B_32, ...], 
+			//		state_root : B_32, block_number : P, block_hash : B_32 ]
+			m_protocolVersion = _r[0].toInt<unsigned>();
+			m_networkId = _r[1].toInt<u256>();
+			m_totalDifficulty = _r[2].toInt<u256>();
+			m_latestHash = _r[3].toHash<h256>();
+			m_genesisHash = _r[4].toHash<h256>();
+			m_snapshotHash = _r[5].toHash<h256>();
+			m_snapshotNumber = _r[6].toInt<u256>();
+
+			clog(NetMessageSummary) << "Status: "
+				<< "protocol version " << m_protocolVersion
+				<< "networkId " << m_networkId
+				<< "genesis hash " << m_genesisHash
+				<< "total difficulty " << m_totalDifficulty
+				<< "latest hash" << m_latestHash
+				<< "snapshot hash" << m_snapshotHash
+				<< "snapshot number" << m_snapshotNumber;
+				setIdle();
+			m_observer->onPeerStatus(dynamic_pointer_cast<WarpPeerCapability>(shared_from_this()));
+			break;
+		}
+		case GetBlockHeadersPacket:
+		{
+			RLPStream s;
+			prep(s, BlockHeadersPacket);
+			sealAndSend(s);
+			break;
+		}
+		case SnapshotManifest:
+		{
+			setIdle();
+			m_observer->onPeerManifest((dynamic_pointer_cast<WarpPeerCapability>(shared_from_this())), _r);
+			break;
+		}
+		case SnapshotData:
+		{
+			setIdle();
+			m_observer->onPeerData((dynamic_pointer_cast<WarpPeerCapability>(shared_from_this())), _r);
+			break;
+		}
+		default:
+			return false;
+		}
+	}
+	catch (Exception const&)
+	{
+		clog(NetWarn) << "Warp Peer causing an Exception:" << boost::current_exception_diagnostic_information() << _r;
+	}
+	catch (std::exception const& _e)
+	{
+		clog(NetWarn) << "Warp Peer causing an exception:" << _e.what() << _r;
+	}
+
+	return true;
+}
+
+/// Validates whether peer is able to communicate with the host, disables peer if not
+bool WarpPeerCapability::validateStatus(h256 const& _genesisHash, std::vector<unsigned> const& _protocolVersions, u256 const& _networkId)
+{
+	std::shared_ptr<SessionFace> s = session();
+	if (!s)
+		return false; // Expired
+
+	if (m_genesisHash != _genesisHash)
+	{
+		disable("Invalid genesis hash");
+		return false;
+	}
+	if (find(_protocolVersions.begin(), _protocolVersions.end(), m_protocolVersion) == _protocolVersions.end())
```

Comment:
I know it's a little less efficient on the average case, but `_procotolVersions.count(m_protocolVersion)` is much better readable and less error-prone.